[Clang] Allow the use of [[gnu::visibility]] with #pragma clang attribute

philnik777 opened the pull request:
I don't see any reason this shouldn't be allowed. AFAICT this is only disabled due to the heuristics used to determine whether it makes sense to allow the use of an attribute with `#pragma clang attribute`.

This allows libc++ to drop `_LIBCPP_HIDE_FROM_ABI` in a lot of places, making the library significantly easier to read.

philnik777 committed:
[Clang] Allow the use of [[gnu::visibility]] with #pragma clang attribute

erichkeane reviewed (approved):
I don't see a reason to not do this, and if it makes the library simpler, even better.  I think you fixed the pre-commit CI now, so LGTM.

AaronBallman commented:
This is missing test coverage for the interesting cases.

I'm a bit worried about how easy it will be to apply this attribute in unexpected places. e.g., it will apply to member functions as well as free functions, will anyone expect that though? It also applies to special functions like `main`, though that's more a question of the attribute and less about the pragma.

What's the behavior going to be in C when you get an implicit function declaration in C89 mode? Does that declaration then pick up this attribute as well? How about other special member functions we define on behalf of the user?

AaronBallman commented:
> @AaronBallman Are any of these concerns specific to `[[gnu::visibility]]`? The exact same concerns seem to apply to any other attribute that applies to functions.

Yes and no. It's a general issue with other attributes, but when we added this pragma, we decided to be conservative with what attributes are allowed. We'd go on a case-by-case basis for whether the attribute will be something users can make sense of using. (For example, we don't support `[[deprecated]]` or `[[maybe_unused]]` etc).

I *think* this may be reasonable, but it would help if there were a significant number of test cases around the edges so we can reason about how usable the functionality is.

philnik777 commented:
> This is missing test coverage for the interesting cases.
> 
> I'm a bit worried about how easy it will be to apply this attribute in unexpected places. e.g., it will apply to member functions as well as free functions, will anyone expect that though?

This is explicitly documented, so I'd say yes. This is how the pragma works for better or worse.

> It also applies to special functions like `main`, though that's more a question of the attribute and less about the pragma.

I'm not sure whether this is a request for a test?

> What's the behavior going to be in C when you get an implicit function declaration in C89 mode? Does that declaration then pick up this attribute as well? How about other special member functions we define on behalf of the user?

I'm not 100% sure what you mean by "define on behalf of users". I've added a test for an explicitly defaulted special member. Note though that using `#pragma clang attribute` is basically like explicitly defaulting the special members and adding the attribute there.

philnik777 committed:
Address comments

philnik777 commented:
> > > This is missing test coverage for the interesting cases.
> > > I'm a bit worried about how easy it will be to apply this attribute in unexpected places. e.g., it will apply to member functions as well as free functions, will anyone expect that though?
> > 
> > 
> > This is explicitly documented, so I'd say yes. This is how the pragma works for better or worse.
> 
> I think you're right, but is that actually helpful in this case? (I'm not super experienced with the visibility attributes, coming from a Windows background -- does it make sense to have "hidden" member functions on a class? If so, then I think the attribute makes sense to support. If not, maybe we want to consider different diagnostics to help users figure out when they've made a potential mistake?)

At least that is exactly what libc++ needs. So yeah, depending on what you want to achieve it can definitely make sense. If someone needs something different we can still add `function(unless(is_member))` or something like that to the matchers.

> CC @compnerd who maybe can help me with my education. :-D
> 
> > > It also applies to special functions like `main`, though that's more a question of the attribute and less about the pragma.
> > 
> > 
> > I'm not sure whether this is a request for a test?
> 
> Yes, along with "maybe we want to consider different diagnostics" if the idea of adding visibility to `main` makes no sense.
> 
> > > What's the behavior going to be in C when you get an implicit function declaration in C89 mode? Does that declaration then pick up this attribute as well? How about other special member functions we define on behalf of the user?
> > 
> > 
> > I'm not 100% sure what you mean by "define on behalf of users". I've added a test for an explicitly defaulted special member.
> 
> That's what I was after.
> 
> > Note though that using `#pragma clang attribute` is basically like explicitly defaulting the special members and adding the attribute there.
> 
> Good to know, I kind of expected that. Does the same happen for an implicit function declaration in C? I can see us defining it with or without the attribute, so it's more a matter of knowing what happens and capturing it in a test so we can be aware if we change that behavior.

I've added a test for implicit declarations in C and that's affected by `#pragma clang attribute`, but it seems to not be affected by `-fvisibility`. Not sure whether this makes any sense. To me implicit declarations don't make any sense in general.

AaronBallman reviewed (approved):
LGTM!

compnerd commented:
> > > This is missing test coverage for the interesting cases.
> > > I'm a bit worried about how easy it will be to apply this attribute in unexpected places. e.g., it will apply to member functions as well as free functions, will anyone expect that though?
> > 
> > 
> > This is explicitly documented, so I'd say yes. This is how the pragma works for better or worse.
> 
> I think you're right, but is that actually helpful in this case? (I'm not super experienced with the visibility attributes, coming from a Windows background -- does it make sense to have "hidden" member functions on a class? If so, then I think the attribute makes sense to support. If not, maybe we want to consider different diagnostics to help users figure out when they've made a potential mistake?)
> 
> CC @compnerd who maybe can help me with my education. :-D

It is possible to apply the hidden visibility to member functions on classes. However, it does tend to complicate your ability to reason about the code IMO (although the same argument holds for `__declspec(dllexport)` on member functions. The attribute simply would prevent the member function from participating in dynamic linking - so any other module would need to define their own copy.
